File: Task2/Java/Main.java
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter text for encryption:");
        String inputText = scanner.nextLine();

        // Convert the input string to bytes using UTF-8 encoding
        // This ensures that multi-byte characters are correctly handled from the start
        byte[] inputBytesPrimitive = inputText.getBytes(StandardCharsets.UTF_8);
        List<Byte> inputBytes = new ArrayList<>();
        for (byte b : inputBytesPrimitive) {
            inputBytes.add(b);
        }

        // Add padding to input bytes
        List<Byte> paddedInputBytes = Padding.addPkcs7Padding(inputBytes);

        // Generate random key (16 bytes for AES-128)
        byte[] aesKey = KeyExpansion.generateRandomKey(16);
        System.out.print("Generated AES Key: ");
        AesCbcCipher.printBytes(aesKey);

        // Generate random initialization vector (16 bytes for IV)
        byte[] initVec = KeyExpansion.generateRandomIV(16);
        System.out.print("Generated Initialization Vector (IV): ");
        AesCbcCipher.printBytes(initVec);

        System.out.println("\n" + "-".repeat(50));
        System.out.println("Starting AES-128 (CBC) Encryption");
        System.out.println("-".repeat(50));

        // Encryption
        AesCbcCipher cbcCipher = new AesCbcCipher();
        List<Byte> encryptedBytes = cbcCipher.encryptCbc(paddedInputBytes, aesKey, initVec);

        System.out.println("\n" + "-".repeat(50));
        System.out.println("Encryption Completed");
        System.out.println("-".repeat(50));
        System.out.print("Encrypted text: ");
        AesCbcCipher.printBytes(toPrimitiveByteArray(encryptedBytes));

        System.out.println("\n" + "-".repeat(50));
        System.out.println("Starting AES-128 (CBC) Decryption");
        System.out.println("-".repeat(50));

        // Decryption
        List<Byte> decryptedBytes = cbcCipher.decryptCbc(encryptedBytes, aesKey, initVec);

        // Remove padding from decrypted bytes
        List<Byte> finalDecryptedBytes = Padding.removePkcs7Padding(decryptedBytes);

        System.out.println("\n" + "-".repeat(50));
        System.out.println("Decryption Completed");
        System.out.println("-".repeat(50));
        System.out.print("Decrypted text: ");
        // Convert the final decrypted bytes to a String using UTF-8 encoding
        System.out.println(new String(toPrimitiveByteArray(finalDecryptedBytes), StandardCharsets.UTF_8));

        scanner.close();
    }

    // Helper to convert List<Byte> to byte[]
    private static byte[] toPrimitiveByteArray(List<Byte> byteList) {
        byte[] byteArray = new byte[byteList.size()];
        for (int i = 0; i < byteList.size(); i++) {
            byteArray[i] = byteList.get(i);
        }
        return byteArray;
    }
}

File: Task2/C++/main.cpp
#include "library.h"

int main() {
	std::srand(
		static_cast<unsigned int>(std::time(nullptr))); // Инициализация генератора случайных чисел

	std::string inputText;
	std::cout << "Введите текст для шифрования (кратный 16 символам): " << std::endl;
	std::getline(std::cin, inputText);
	/*size_t length = inputText.length();
	if (static_cast<int>(length) % 16 != 0)
	inputText.append(length, '\0');*/

	std::vector<uint8_t> inputBytes(inputText.begin(), inputText.end());

	// Добавление отступа к входным байтам
	std::vector<uint8_t> paddedInputBytes = addPkcs7Padding(inputBytes);

	// Генерация случайного ключа и запись его в файл
	std::vector<uint8_t> aesKey = generateRandomKey(16); // 16 байт для AES-128
	printBytes(aesKey);

	// Генерация случайного вектора инициализации
	std::vector<uint8_t> initVec = generateRandomIV(16); // 16 байт для IV
	std::cout << "Сгенерированный вектор инициализации (IV): ";
	printBytes(initVec);

	std::cout << std::endl << std::string(50, '-') << std::endl;
	std::cout << "Начало шифрования AES-128 (CBC)" << std::endl;
	std::cout << std::string(50, '-') << std::endl;

	// Шифрование
	std::vector<uint8_t> encryptedBytes = aes128EncryptCbc(paddedInputBytes, aesKey, initVec);

	std::cout << std::endl << std::string(50, '-') << std::endl;
	std::cout << "Шифрование завершено" << std::endl;
	std::cout << std::string(50, '-') << std::endl;
	std::cout << "Зашифрованный текст: ";
	printBytes(encryptedBytes);

	std::cout << std::endl << std::string(50, '-') << std::endl;
	std::cout << "Начало расшифрования AES-128 (CBC)" << std::endl;
	std::cout << std::string(50, '-') << std::endl;

	// Расшифрование
	std::vector<uint8_t> decryptedBytes = aes128DecryptCbc(encryptedBytes, aesKey, initVec);

	// Удаление отступа из расшифрованных байтов
	std::vector<uint8_t> finalDecryptedBytes = removePkcs7Padding(decryptedBytes);

	std::cout << std::endl << std::string(50, '-') << std::endl;
	std::cout << "Расшифрование завершено" << std::endl;
	std::cout << std::string(50, '-') << std::endl;
	std::cout << "Расшифрованный текст: " << std::endl;
	for (uint8_t byte : finalDecryptedBytes) {
		std::cout << static_cast<char>(byte);
	}
	std::cout << std::endl;

	return 0;
}
